### ISSUE TITLE
Vanishing Point filter (tool for real perspective for layers)

### ISSUE THREAD

**Issue by MichalSZZ** [photopea/photopea#2355]
Hi, this tool sometimes help a lot in adding the perspective to design easliy.
https://www.youtube.com/watch?v=76A1cytMYGs in just 2 minuts this video explain everything :)

this tool have possibility to work with stamp tool or brush tool in perspective too
https://www.youtube.com/watch?v=3ZGBhWh_JqM

**Comment by photopea** [comment]
Wow, this filter requires a brand new user interface (new toolbar etc).

Do you know that you can drag corners independently with a Free Transform, if you hold Ctrl? See perspective transform here: https://www.photopea.com/learn/free-transform

**Comment by MichalSZZ** [comment]
> Do you know that you can drag corners independently with a Free Transform, if you hold Ctrl? See perspective transform here: https://www.photopea.com/learn/free-transform

:) yes - I know that :) If you check closed issues here - that was my suggestion too (for more complicated perspectiv hack with shortcuts) https://github.com/photopea/photopea/issues/769

that tool from this filter can save time in work - look for more videos on youtube (for example - I can paste layer with shadow and curve it automaticaly on the ground and wall)

**Comment by abdullah20401** [comment]
I would also love to have this tool added into photopea.

**Comment by Bastienre4** [comment]
+1 :)

**Comment by ofekkazes** [comment]
+1
Please

**Comment by CWBudde** [comment]
> Wow, this filter requires a brand new user interface (new toolbar etc).

Indeed, it requires a deeper integration, as it might look in the first place. But I think every required tools for the transformations are already there.

What needs to be done next is to set up a roadmap, eventually with dependencies, to see how this can be implemented in the near future.

> Do you know that you can drag corners independently with a Free Transform, if you hold Ctrl? See perspective transform here: https://www.photopea.com/learn/free-transform

It's already working with free-transform, but it is a lot of manual work. Especially if you move the item around on a plane with a potential (nonuniform) scaling. This is much easier if all done by the software.

Overall, it's probably not the killer feature, but it's also not rocket science to implement.

**Comment by photopea** [comment]
> not rocket science to implement.

Could you help us implement it? :D

**Comment by MichalSZZ** [comment]
In the last month I used one function from Photoshop Vanishing Point. That was 'exporting the grid mesh as a layer'. In photoshop this function is hidden.

But two functions most used by me are:
1. Using stamp tool in perspective
2. Pasting layer from clipboard like here https://youtube.com/shorts/SCl6VMsIikU?feature=share

**Comment by photopea** [comment]
I have implemented the Vanishing Point filter. It is under Filter - Vanishing Point.

It works similarly to the one in Adobe Photoshop. It is fully compatible with the PSD format. Could you test it out and let me know about possible bugs?

**Comment by MichalSZZ** [comment]
Wow. Thanks. Many things to test out.
I will write it in parts. The list isn't ordered in any way (priority or something like that)

1. After making the first perspective grid it should to have grid and active control points visible (not after clicking on it)
2. I can't set the Brush Size in the Stamp Tool with a keyboard Shortcut [ and ]. Same thing in Brush Tool.
3. As above but for hardness with { and }
4. Sometimes I am able to do a triangle grid (the red one). When I find out how to reproduce that I will tell you how to achieve it (but it happened to me few times)
5. When I try to delete created grid with "del" or "backspace" I have the prompt "Close the current window first"
6. When I try to back one step in creating the plane - backspace doesn't work (should work like in polygonal lasso tool)
7. When I try to make a new perspective grid and hit "esc" - it doesn't cancel the operation - it should do that (but in Photoshop it doesn't too - I don't like it :) ). In Polygonal Lasso Tool in PS and Photoshop, ESC cancels the whole process. The same in the Path tool.
8. In editing grid's plane mode, the cursor icon for transforming the middle top and bottom point should be an up-down double arrow.
9. It shouldn't be possible to do crossing lines in a plane (polygonal with 4 sides that looks like 8 - if you know what I mean).
10. Same thing with a plane that has one of the angle bigger than 90 degrees.

will be more soon :)

**Comment by MichalSZZ** [comment]
another part:

1. Selections that were created in the Vanishing Point using its Selection Tool -Shouldn't look like the plane (should look like selection with dashed line)
2. The Brush in a Stamp Tool doesn't have a live view of the sampled area
3. When I paste something from the clipboard (that was copied in Photopea with ctrl+c) it pastes upside-down and in the left bottom grid (that I made bigger than canvas) 
4. When I zoom out and use the hand tool to move the view. The gray background ends and I am able to see the chessboard
5. I don't know how to change the position of the corner after making a new plane (made by extending the first plane to the another axis). I can move the middle points but not the corners. But it is very complicated - moving that point moves all planes joined with that plane)

**Comment by photopea** [comment]
I have implemented everything except of the live preview of the stamp tool. That would be pretty hard to make.

I do not understand the last request. If you have just one plane (and nothing attached to it), you can drag corners with any of the first two tools. With more than one related planes, you can not move corners, as it would ruin the system. The "angle" of the plane in space is fixed now, you can only extend it to the sides.

**Comment by photopea** [comment]
I have implemented the brush preview for the Brush tool and the Stamp tool inside the Vanishing point filter.

**Comment by MichalSZZ** [comment]
@photopea I understand. That how it works in Photoshop. But it is not so important.

For all your fixes - huge respect! 
I couldn't test anything more, because my daughter was sick and I didn't have time.

**Comment by photopea** [comment]
@MichalSZZ the brush preview works now. Hope your daughter gets better!

**Comment by photopea** [comment]
I wrote a tutorial for this filter: https://www.photopea.com/learn/vanishing-point
